// farce/lib/BrowserProtocol.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _off = require('dom-helpers/events/off');

var _off2 = _interopRequireDefault(_off);

var _on = require('dom-helpers/events/on');

var _on2 = _interopRequireDefault(_on);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

var _createPath = require('./utils/createPath');

var _createPath2 = _interopRequireDefault(_createPath);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var BrowserProtocol = function () {
  function BrowserProtocol() {
    _classCallCheck(this, BrowserProtocol);

    this._keyPrefix = Math.random().toString(36).slice(2, 8);
    this._keyIndex = 0;

    this._index = null;
  }

  BrowserProtocol.prototype.init = function init() {
    var _window$location = window.location,
        pathname = _window$location.pathname,
        search = _window$location.search,
        hash = _window$location.hash;

    var _ref = window.history.state || {},
        key = _ref.key,
        _ref$index = _ref.index,
        index = _ref$index === undefined ? 0 : _ref$index,
        state = _ref.state;

    var delta = this._index != null ? index - this._index : 0;
    this._index = index;

    return {
      action: 'POP',
      pathname: pathname,
      search: search,
      hash: hash,
      key: key,
      index: index,
      delta: delta,
      state: state
    };
  };

  BrowserProtocol.prototype.subscribe = function subscribe(listener) {
    var _this = this;

    var onPopState = function onPopState() {
      listener(_this.init());
    };

    // TODO: On most versions of IE, we need a hashChange listener for hash-
    // only changes.

    (0, _on2.default)(window, 'popstate', onPopState);
    return function () {
      return (0, _off2.default)(window, 'popstate', onPopState);
    };
  };

  BrowserProtocol.prototype.transition = function transition(location) {
    var action = location.action,
        state = location.state;


    var push = action === 'PUSH';
    !(push || action === 'REPLACE') ? process.env.NODE_ENV !== 'production' ? (0, _invariant2.default)(false, 'Unrecognized browser protocol action ' + action + '.') : (0, _invariant2.default)(false) : void 0;

    var delta = push ? 1 : 0;
    var extraState = this._createExtraState(delta);

    var browserState = _extends({ state: state }, extraState);
    var path = (0, _createPath2.default)(location);

    if (push) {
      window.history.pushState(browserState, null, path);
    } else {
      window.history.replaceState(browserState, null, path);
    }

    return _extends({}, location, extraState, { delta: delta });
  };

  BrowserProtocol.prototype.go = function go(delta) {
    window.history.go(delta);
  };

  BrowserProtocol.prototype.createHref = function createHref(location) {
    return (0, _createPath2.default)(location);
  };

  BrowserProtocol.prototype._createExtraState = function _createExtraState(delta) {
    var keyIndex = this._keyIndex++;
    this._index += delta;

    return {
      key: this._keyPrefix + ':' + keyIndex.toString(36),
      index: this._index
    };
  };

  return BrowserProtocol;
}();

exports.default = BrowserProtocol;
module.exports = exports['default'];
